### PR TITLE
MM-34712 shared channels: sync user profile image

### PR DESCRIPTION
**Pull request by wiggin77**
#### Summary
This PR adds support for synchronizing user's profile image for shared channels.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-34712

#### Release Note
```release-note
NONE
```

### CODE REVIEW

**Review by Xxoopp20115 — APPROVED** [review]

**Review by Xxoopp20115 — COMMENTED** [review]
A

**Review by Xxoopp20115 — CHANGES_REQUESTED** [review]
A

File: services/remotecluster/sendprofileImage.go [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+	u, err := url.Parse(task.rc.SiteURL)
+	if err != nil {
+		return fmt.Errorf("invalid siteURL while sending file to remote %s: %w", task.rc.RemoteId, err)
+	}
+	u.Path = path.Join(u.Path, model.API_URL_SUFFIX, "remotecluster", task.userID, "image")
+
+	body := &bytes.Buffer{}
+	writer := multipart.NewWriter(body)
+
+	part, err := writer.CreateFormFile("image", "profile.png")
```
Comment (agnivade) [inline services/remotecluster/sendprofileImage.go:118]: So this seems like we are reading the entire image, writing it to a byte buffer, and then reading from the buffer again to send it to the remote. Perhaps out of scope for this PR, but I am wondering if we can directly stream the already read image directly to the POST request?

It's another issue though that we don't have streaming read APIs from the app layer.

File: services/remotecluster/sendprofileImage.go [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+		provider: provider,
+		f:        f,
+	}
+	return rcs.enqueueTask(ctx, rc.RemoteId, task)
+}
+
+// sendProfileImage is called when a sendProfileImageTask is popped from the send channel.
+func (rcs *Service) sendProfileImage(task sendProfileImageTask) {
+	// Ensure a panic from the callback does not exit the goroutine.
+	defer func() {
```
Comment (agnivade) [inline services/remotecluster/sendprofileImage.go:58]: Are we actually panicking from somewhere? If not, then I am wondering why are we doing this only for `sendProfileImage`? It's usually a good practice to let the entire thing crash for a panic.

File: services/remotecluster/sendprofileImage.go [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+		response.Status = ResponseStatusOK
+	}
+
+	// If callback provided then call it with the results.
+	if task.f != nil {
+		task.f(task.userID, task.rc, &response, err)
+	}
+}
+
+func (rcs *Service) sendProfileImageToRemote(timeout time.Duration, task sendProfileImageTask) error {
```
Comment (agnivade) [inline services/remotecluster/sendprofileImage.go:93]: Would love to have a test for this.

File: services/remotecluster/sendprofileImage.go (hunk 2 above)
Comment (wiggin77) [reply services/remotecluster/sendprofileImage.go:58]: This is done for all three of the sendXXX operations that the remote cluster service supports. It comes from a habit when allowing callbacks from arbitrary places to not trust those callbacks not to crash.  I guess they can be removed.  I could also isolate the panic handler to only guard the callback.  ... ... ...  I think I'll just remove them.

File: services/remotecluster/sendprofileImage.go (hunk 1 above)
Comment (wiggin77) [reply services/remotecluster/sendprofileImage.go:118]: The app layer API to fetch the image returns a []byte of the entire image.  We then need to convert that to a multipart request for the POST.  I could just put the raw binary in the payload and adjust the receiving end to expect that, but that would deviate from how we're doing things with other resource uploads.

If the app layer API provided a reader I could skip the `bytes.NewBuffer` creation.  Profile images are limited to 128x128 so I don't think it is worth it as this stage.  If we fix it we should do it for all resource reading/writing, especially the bigger assets.

File: services/remotecluster/sendprofileImage.go (hunk 2 above)
Comment (wiggin77) [reply services/remotecluster/sendprofileImage.go]: Done

File: services/remotecluster/sendprofileImage.go (hunk 3 above)
Comment (wiggin77) [reply services/remotecluster/sendprofileImage.go:82]: Unit test added.

**Review by agnivade — APPROVED** [review]

File: services/remotecluster/sendprofileImage.go [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+	if _, err = io.Copy(part, bytes.NewBuffer(img)); err != nil {
+		return err
+	}
+
+	if err = writer.Close(); err != nil {
+		return err
+	}
+
+	req, err := http.NewRequest("POST", u.String(), bytes.NewReader(body.Bytes()))
```
Comment (streamer45) [inline services/remotecluster/sendprofileImage.go:120]:
```suggestion
	req, err := http.NewRequest("POST", u.String(), body)
```

File: services/remotecluster/sendprofileImage_test.go [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+		}
+	}
+
+	buf := &bytes.Buffer{}
+	png.Encode(buf, img)
+
+	return buf.Bytes(), true, nil
+}
+
+type Flag struct {
```
Comment (streamer45) [inline services/remotecluster/sendprofileImage_test.go:172]: Does this (and related methods) need to be exported?

File: services/remotecluster/sendprofileImage_test.go [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+		for y := 0; y < imageHeight; y++ {
+			img.Set(x, y, red)
+		}
+	}
+
+	buf := &bytes.Buffer{}
+	png.Encode(buf, img)
+
+	return buf.Bytes(), true, nil
+}
```
Comment (streamer45) [inline services/remotecluster/sendprofileImage_test.go:170]: Usually it's just simpler to read some test asset from `tests/` but I like your initiative.

**Review by streamer45 — APPROVED** [review]
Looks good :+1:  Only left non-blocking suggestions.

File: services/remotecluster/sendprofileImage_test.go (hunk 5 above)
Comment (wiggin77) [reply services/remotecluster/sendprofileImage_test.go]: No, and fixed.

**Review by furqanmlk — APPROVED** [review]
**Name Collision Passed** ✅ 
image

Sync users profile image and Sync custom status passed ✅ 
There are few issues with these feature, will be addressed separately

**Pull request merged**